File: MCRRTImp/src/test/java/infrastructure/ui/GUIBase.java
package infrastructure.ui;

import javafx.application.Application;
import javafx.application.Platform;
import javafx.scene.Scene;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Pane;
import javafx.stage.Stage;

public abstract class GUIBase extends Application {

    protected int height = 2000;
    protected int width = 2000;
    protected double cycleSleepTime = 1 / 60.0;
    private Pencil userPencil;
    private GraphicsContext pencil;

    protected double scrollZoomBase = 1;
    private double lastMouseX = -1;
    private double lastMouseY = -1;
    protected double scaleBase = 1;

    protected double shiftX = 1;
    protected double shiftY = 1;

    protected abstract void draw(Pencil pencil);


    protected void initializeComponents(Stage primaryStage, Scene scene, Pane root, Canvas canvas){
        primaryStage.widthProperty().addListener((ob, ov, nv) -> {
            this.width = (int)(double) nv;
            canvas.setWidth((double) nv);

        });
        primaryStage.heightProperty().addListener((ob, ov, nv) -> {
            this.height = (int)(double) nv;
            canvas.setHeight((double) nv);

        });
        primaryStage.widthProperty().addListener((ob, ov, nv) -> scaleBase = ((double) nv * scrollZoomBase) / width);
        canvas.setOnScroll(event -> {
            double deltaY = event.getDeltaY();
            if (deltaY < 0) {
                scrollZoomBase = scrollZoomBase > 0.2 ? scrollZoomBase - 0.1 : 0.2;
            } else if (deltaY > 0) {
                scrollZoomBase = scrollZoomBase < 100 ? scrollZoomBase + 0.1 : 100;
            }
            scaleBase = (scene.getWidth() * scrollZoomBase) / width;
        });
        canvas.setOnMouseDragged(event -> {
            double mouseX = event.getX();
            double mouseY = event.getY();
            if (lastMouseX == -1) {
                lastMouseX = mouseX;
                lastMouseY = mouseY;
                return;
            }
            double deltaX = mouseX - lastMouseX;
            double deltaY = mouseY - lastMouseY;
            shiftX += deltaX;
            shiftY += deltaY;
            lastMouseX = mouseX;
            lastMouseY = mouseY;
        });
        canvas.setOnMouseMoved(event -> {
            lastMouseX = event.getX();
            lastMouseY = event.getY();
        });
        primaryStage.heightProperty().addListener((ob, ov, nv) -> scaleBase = (double) nv / height);
        primaryStage.setOnCloseRequest(event -> System.exit(0));
    }

    private void thread(){
        while (true) {
            try {
                Thread.sleep((long) (1000 * cycleSleepTime));
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            Platform.runLater(() -> {
                pencil.clearRect(0, 0, width, height);
                userPencil.scale(scaleBase).pixelOffset(shiftX, shiftY);
                draw(userPencil);
            });
        }
    }

    @Override
    public final void start(Stage primaryStage) {
        HBox root = new HBox();
        Canvas canvas = new Canvas(width, height);
        root.getChildren().add(canvas);
        Scene scene = new Scene(root, width, height);
        primaryStage.setScene(scene);
        pencil = canvas.getGraphicsContext2D();
        userPencil = new Pencil(pencil);
        initializeComponents(primaryStage,scene , root, canvas);
        Thread redrawTrigger = new Thread(this::thread);
        redrawTrigger.start();
        primaryStage.show();
    }

}

File: MCRRTImp/src/test/kotlin/infrastructure/ui/GUIBase.kt
package infrastructure.ui

import infrastructure.Vector2BasedImp.Vector2
import javafx.application.Application
import javafx.application.Platform
import javafx.scene.Scene
import javafx.scene.canvas.Canvas
import javafx.scene.canvas.GraphicsContext
import javafx.scene.layout.HBox
import javafx.scene.layout.Pane
import javafx.scene.paint.Color
import javafx.stage.Stage

abstract class GUIBase : Application() {

    protected var height = 2000
    protected var width = 2000
    protected var cycleSleepTime = 1 / 60.0
    private lateinit var userPencil: Pencil
    private lateinit var pencil: GraphicsContext

    protected var scrollZoomBase = 1.0
    private var lastMouseX = -1.0
    private var lastMouseY = -1.0
    protected var scaleBase = 1.0

    protected var shiftX = 1.0
    protected var shiftY = 1.0

    protected abstract fun draw(pencil: Pencil)


    protected open fun initializeComponents(primaryStage: Stage, scene: Scene, root: Pane, canvas: Canvas) {
        primaryStage.widthProperty().addListener { ob, ov, nv ->
            this.width = (nv as Double).toInt()
            canvas.width = nv

        }
        primaryStage.heightProperty().addListener { ob, ov, nv ->
            this.height = (nv as Double).toInt()
            canvas.height = nv

        }
        primaryStage.widthProperty().addListener { ob, ov, nv -> scaleBase = nv as Double * scrollZoomBase / width }
        canvas.setOnScroll { event ->
            val deltaY = event.deltaY
            if (deltaY < 0) {
                scrollZoomBase = if (scrollZoomBase > 0.2) scrollZoomBase - 0.1 else 0.2
            } else if (deltaY > 0) {
                scrollZoomBase = if (scrollZoomBase < 100) scrollZoomBase + 0.1 else 100.0
            }
            scaleBase = scene.width * scrollZoomBase / width
        }
        canvas.setOnMouseDragged { event ->
            val mouseX = event.x
            val mouseY = event.y
            if (lastMouseX == -1.0) {
                lastMouseX = mouseX
                lastMouseY = mouseY
                return@setOnMouseDragged
            }
            val deltaX = mouseX - lastMouseX
            val deltaY = mouseY - lastMouseY
            shiftX += deltaX
            shiftY += deltaY
            lastMouseX = mouseX
            lastMouseY = mouseY
        }
        canvas.setOnMouseMoved { event ->
            lastMouseX = event.x
            lastMouseY = event.y
        }
        primaryStage.heightProperty().addListener { ob, ov, nv -> scaleBase = nv as Double / height }
        primaryStage.setOnCloseRequest { event -> System.exit(0) }
    }

    private fun thread() {
        while (true) {
            try {
                Thread.sleep((1000 * cycleSleepTime).toLong())
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }

            Platform.runLater {
                pencil.clearRect(0.0, 0.0, width.toDouble(), height.toDouble())
                userPencil.scale(scaleBase).pixelOffset(shiftX, shiftY)
                userPencil.filled().color(Color.BLACK).rect(Vector2(width / 2.0, height / 2.0), Vector2(width.toDouble(), height.toDouble()))
                draw(userPencil)
            }
        }
    }

    override fun start(primaryStage: Stage) {
        val root = HBox()
        val canvas = Canvas(width.toDouble(), height.toDouble())
        root.children.add(canvas)
        val scene = Scene(root, width.toDouble(), height.toDouble())
        primaryStage.scene = scene
        pencil = canvas.graphicsContext2D
        userPencil = Pencil(pencil)
        initializeComponents(primaryStage, scene, root, canvas)
        val redrawTrigger = Thread(Runnable { this.thread() })
        redrawTrigger.start()
        primaryStage.show()
    }

}
